dbms/src/Storages/MergeTree/RPNBuilder.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include <Common/typeid_cast.h>
+#include <Core/Block.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/ExpressionAnalyzer.h>
+#include <Parsers/ASTExpressionList.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTFunction.h>
+#include <Storages/SelectQueryInfo.h>
+
+namespace DB
+{
+
+
+/** Calculate expressions, that depend only on constants.
+  * For index to work when something like "WHERE Date = toDate(now())" is written.
+  */
+static Block getBlockWithConstants(
+    const ASTPtr & query, const SyntaxAnalyzerResultPtr & syntax_analyzer_result, const Context & context)
+{
+    Block result
+    {
+        { DataTypeUInt8().createColumnConstWithDefaultValue(1), std::make_shared<DataTypeUInt8>(), "_dummy" }
+    };
+
+    const auto expr_for_constant_folding = ExpressionAnalyzer(query, syntax_analyzer_result, context).getConstActions();
+
+    expr_for_constant_folding->execute(result);
+
+    return result;
+}
+
+
+/// Builds reverse polish notation
+template <typename RPNElement>
+class RPNBuilder
```

Comment:
Can be also used in KeyCondition, but it fails performance tests. (For BloomFilterIndex I have not noticed any difference in performance between copy-paste functions and RPNBuilder)